### ISSUE TITLE
Some test-case lockup

### ISSUE THREAD

**Issue by kas1e** [ptitSeb/gl4es#48]
There is test case i do check for now on amigaos4:
http://kas1e.mikendezign.com/aos4/gl4es/tests/drawbuffer_gl4es.c

Once i run it, after few loops of DrawBox() , it crashes.

It works through when pure opengl used (some old 1.1 version which we have), but can be that bug in the code as well.   There is only defined 6 normals for 8 vertices, but probably that should't cause crash ?

@ptitSeb
Can you check plz that GL code on your own (with SDL or something) , to see if that code works at all for on gl4es, so we can then know if it amigaos4 only issue or not.

Thanks!

**Comment by ptitSeb** [comment]
@kas1e I have tested on the Pandora, and I have no lockup. I see an oversized cube, witch each face  with is own color, rotating.

Did you locate where the crash occurs?

Also: the Normal is not an issue. You have not defined a NormalArray here, but you are just defining the Normal 1 time per face., with a simple `glNormal3fv`. The code is correct.

**Comment by kas1e** [comment]
>  I have tested on the Pandora, and I have no lockup. I see an oversized cube, witch each face with its own color, rotating.

Yes, that how it should be (oversized just because i comment out 2 GLU functions to be able to build without).

> Did you locate where the crash occurs?

Its hard, as it lockup whole machine and i can't normally debug it , but all i can see, is that when i put prinfs after and before each function, i can see how drawbox() called for about 10 times (and code from executes), and then, crashes on glNormal3fv().

Can it be issues with the same uniforms and stuff which we workoround yesterday ?

**Comment by ptitSeb** [comment]
Mmm, you are not in a drawing loop when you call `glNormal3fv(..)` so not much is done with the info. `fpe_glNormal3f` should be called, which does nothing... (in fact, only things hapening is storing current normal in glstate for later use, in the drawing shader because light is enabled and normal is needed).

It can be something odd with Uniform/Attrib yes.

You should activate some debug info in gl4es, and do only 2 loop and exit, to avoid freeze and get access to logs.

**Comment by kas1e** [comment]
> It can be something odd with Uniform/Attrib yes.

Author of our driver already checking it, probably if all going well will have fixed version today (and so we for first can remove workoround, and for second can be sure if it because of that crashes in that example, or not).

> You should activate some debug info in gl4es, and do only 2 loop and exit, to avoid freeze and get access to logs.

Indeed !  Thansk for idea :)

**Comment by kas1e** [comment]
@ptitSeb
I recieve fixed version of ogles2.library , so we can remove workoround for amigaos4 in the gl/program.c

Through, it didn't fish crash about which that ticket, so i will follow now your suggestion and will try to enable as much debug as possible

**Comment by ptitSeb** [comment]
Done, workaround removed.

You should put some traces in `fpe.c`, `shaderconv.c` and `program.c` at least, that should give a good idea on what is going on.

**Comment by kas1e** [comment]
So, current source of that test case which crashes: http://kas1e.mikendezign.com/aos4/gl4es/tests/gl4es_crash.c

And to be on safe side from begining, i enable debug everywhere:

gl/buffers.c
gl/fpe.c
gl/framebuffer.c
gl/matrix.c
gl/program.c
gl/shader.c
gl/shaderconv.c
gl/texture.c
gl/uniform.c
gl/vertexattrib.c

And just do it like:

  for (i = 0; i < 6; i++) {
       // exit after 3 !
	if  (i>3) { exit(0);};
	
	glEnableClientState(GL_VERTEX_ARRAY);
    glNormal3fv(&n[i][0]);
    glVertexPointer(3, GL_FLOAT, 0, v);
    glColor3fv(&colors[i][0]);
	
    glDrawElements(GL_QUADS, 4, GL_UNSIGNED_INT, faces[i]);
  }
}

full source here: http://kas1e.mikendezign.com/aos4/gl4es/tests/gl4es_crash_try_to_debug.c

And the debug log is:  http://kas1e.mikendezign.com/aos4/gl4es/tests/gl4es_crash_try_to_debug.txt

After program exit , after 2 or 3 seconds whole lockup.

Its like some heavy trashing of some memory somewhere ..

**Comment by kas1e** [comment]
To be on safe side from begining, i enable debug everywhere:

gl/buffers.c
gl/fpe.c
gl/framebuffer.c
gl/matrix.c
gl/program.c
gl/shader.c
gl/shaderconv.c
gl/texture.c
gl/uniform.c
gl/vertexattrib.c

And just do it like:

for (i = 0; i < 6; i++) {
// exit after 3 !
if (i>3) { exit(0);};

glEnableClientState(GL_VERTEX_ARRAY);
glNormal3fv(&n[i][0]);
glVertexPointer(3, GL_FLOAT, 0, v);
glColor3fv(&colors[i][0]);

glDrawElements(GL_QUADS, 4, GL_UNSIGNED_INT, faces[i]);
}
}

full source here: http://kas1e.mikendezign.com/aos4/gl4es/tests/gl4es_crash_try_to_debug.c

And the debug log is: http://kas1e.mikendezign.com/aos4/gl4es/tests/gl4es_crash_try_to_debug.txt

After program exit , after 2 or 3 seconds whole lockup.

Its like some heavy trashing of some memory somewhere ..

**Comment by ptitSeb** [comment]
For reference, the shader used is:
Vertex:
```
#version 100
precision mediump float;
precision mediump int;
uniform highp mat4 _gl4es_ModelViewProjectionMatrix;
uniform highp mat3 _gl4es_NormalMatrix;
attribute highp vec4 _gl4es_Vertex;
attribute lowp vec4 _gl4es_Color;
attribute highp vec3 _gl4es_Normal;
struct _gl4es_LightModelParameters {
  vec4 ambient;
};
uniform _gl4es_LightModelParameters _gl4es_LightModel;
struct _gl4es_MaterialParameters
{
   vec4 emission;
   vec4 ambient;
   vec4 diffuse;
   vec4 specular;
   float shininess;
};
uniform _gl4es_MaterialParameters _gl4es_FrontMaterial;
uniform _gl4es_MaterialParameters _gl4es_BackMaterial;
// FPE_Shader generated
// ** Vertex Shader **
// ligthting=1 (twosided=0, separate=0, color_material=1)
// secondary=0, planes=000000
// texture=00000000, point=0
varying vec4 Color;
struct _gl4es_FPELightSourceParameters1
{
   highp vec4 ambient;
   highp vec4 diffuse;
   highp vec4 specular;
   highp vec4 position;
   highp vec3 spotDirection;
   highp float spotExponent;
   highp float spotCosCutoff;
   highp float constantAttenuation;
   highp float linearAttenuation;
   highp float quadraticAttenuation;
};
struct _gl4es_FPELightSourceParameters0
{
   highp vec4 ambient;
   highp vec4 diffuse;
   highp vec4 specular;
   highp vec4 position;
   highp vec3 spotDirection;
   highp float spotExponent;
   highp float spotCosCutoff;
};
struct _gl4es_LightProducts
{
   highp vec4 ambient;
   highp vec4 diffuse;
   highp vec4 specular;
};
uniform _gl4es_FPELightSourceParameters0 _gl4es_LightSource_0;
uniform _gl4es_LightProducts _gl4es_FrontLightProduct_0;

void main() {
vec3 normal = _gl4es_NormalMatrix * _gl4es_Normal;
gl_Position = _gl4es_ModelViewProjectionMatrix * _gl4es_Vertex;
// ColorMaterial On/Off=1 Front = 0 Back = 0
Color = _gl4es_FrontMaterial.emission;
Color += _gl4es_Color*_gl4es_LightModel.ambient;
highp float att;
highp float spot;
highp vec3 VP;
highp float lVP;
highp float nVP;
highp vec3 aa,dd,ss;
highp vec3 hi;
// light 0 on, light_direction=0, light_cutoff180=0
att = 1.0;
VP = normalize(_gl4es_LightSource_0.position.xyz);
aa = _gl4es_Color.xyz * _gl4es_LightSource_0.ambient.xyz;
nVP = dot(normal, VP);
dd = (nVP>0.)?(nVP * _gl4es_Color.xyz * _gl4es_LightSource_0.diffuse.xyz):vec3(0.);
hi = normalize(VP + vec3(0., 0., 1.));
lVP = dot(normal, hi);
ss = (nVP>0. && lVP>0.)?(_gl4es_FrontLightProduct_0.specular.xyz):vec3(0.);
Color.rgb += att*(aa+dd+ss);
// end of light 0
Color.a = _gl4es_Color.a;
Color.rgb = clamp(Color.rgb, 0., 1.);
// texturing
}
```
Fragment:
```
#version 100
precision mediump float;
precision mediump int;
// FPE_Shader generated
// ** Fragment Shader **
// lighting=1, alpha=0, secondary=0, planes=000000, texture=00000000, texformat=00000000 point=0
varying vec4 Color;
void main() {
vec4 fColor = Color;
gl_FragColor = fColor;
}
```

**Comment by ptitSeb** [comment]
There are still some issue with Uniform :(

The Mat3 and Mat4 are fixed! 
But there is an issue with struct uniform now.
For example, the `_gl4es_FrontMaterial` uniform is of type `struct _gl4es_MaterialParameters `
So, the glGetUniform should return all individual struct element, not the struct itself.
AmigaOS4 OGLES2 return:
```
uniform #2 : "_gl4es_FrontMaterial" (builtin)  type=GL_POINTS size=1
```
(note: `type=GL_POINTS` means `type=0x0000`)
it should return:
```
 uniform #3 : "_gl4es_FrontMaterial.emission" (builtin)  type=GL_FLOAT_VEC4 size=1
 uniform #4 : "_gl4es_FrontMaterial.ambient" (builtin)  type=GL_FLOAT_VEC4 size=1
 uniform #5 : "_gl4es_FrontMaterial.diffuse" (builtin)  type=GL_FLOAT_VEC4 size=1
 uniform #6 : "_gl4es_FrontMaterial.specular" (builtin)  type=GL_FLOAT_VEC4 size=1
 uniform #7 : "_gl4es_FrontMaterial.shininess" (builtin)  type=GL_FLOAT size=1
```

I don't see any easy workaround here. Can you open a new bug? Hopefully it's easlily fixable?

**Comment by ptitSeb** [comment]
In fact, using a struct, you really create as many different uniform as there is struct elements...

**Comment by kas1e** [comment]
Tested shaders : they both ok.

As for issue you find, will notice author of driver now

Thanks for help !

**Comment by ptitSeb** [comment]
In gl4es side, that means the light and material parameters are never updated, so there are at 0. the shaders then try to compute stuff with that, it may not be good...

**Comment by kas1e** [comment]
Btw, while experimenting with it, what you think is it worth adding something like -DDEBUG_ALL=1 so, once someone will need to debug everything , it will be enough to recompile gl4es with that ? Just as idea of course, maybe uncommenting per file its easy for debug purporses..

**Comment by ptitSeb** [comment]
I prefer to not have a test at the beggining of each function (like `if(globals4es.debug) printf(...);`, because it's a waste of cpu cycle for most case,  were the `#define` is more efficiant. Of course, that means recompiling, but I prefer speed...

**Comment by ptitSeb** [comment]
Aaah, I missed your point

If you want to activate all debug, simply add `-DDEBUG=1` and you'll have a super chatty gl4es version!

**Comment by kas1e** [comment]
Yeah, that what i mean, thanks for point out !

Author of ogles2 driver works on "struct" stuff, will wait then once he done.

Probably, until structs is fixed, we can't go futher anyway, as in any shader you generate strcuts will be used probably ?

**Comment by ptitSeb** [comment]
Yeah, I don't use array, but I do use struct a lot. We need to wait for a fix, as removing struct is a lot of work.

**Comment by kas1e** [comment]
Yeah, sure we need to fix it on our side , as it will make gles2 driver be better, and will avoid unnecessary hassle in gl4es. Hope today/tomorrow author will be done with, so we can continue tests :)

**Comment by ptitSeb** [comment]
Ok. (in the mean time, take a look at my github account, there are a few games that can be worth porting on AmigaOS4 ;) )

**Comment by kas1e** [comment]
Yeah, will try them all :)

Btw, is it possible add to cmake in the amigaos4 area, to set compile to be "ppc-amigaos-gcc" ? I mean something like:

# AmigaOS4
if(AMIGAOS4)
    set_compier_gcc("ppc-amigaos-gcc")
    set_compiler_g++("ppc-amigao-g++")
    set_compiler_ar("ppc-amigaos-ar")
    set_compiler_ranlib("ppc-amigaos-ranlib")
    add_definitions(-DAMIGAOS4)
    set(NOX11 ON)
    set(NOEGL ON)
endif()

**Comment by ptitSeb** [comment]
Sure I can.

I also see for an option to build as static lib or dynamic

**Comment by ptitSeb** [comment]
Done. Commit 271e72898558598338aa7c63073092ed72928e45 force the compiler for AMIGAOS, and there is a new option if the cmake "STATICLIB" to build gl4es as a static lib.
I haven't forced AMIGAOS4 to use STATICLIB, but I can if it make no sense to build as a dynamic lib.

**Comment by kas1e** [comment]
Checked, seems something wrong ..

cmake . -DAMIGAOS4=1; make GL

give me :

-- The C compiler identification is unknown
-- The CXX compiler identification is unknown

And in log i can see it still tries to use /usr/bin/cc  

But as i cross-compile (on cygwin), maybe that make difference ? ppc-amigaos-gcc and co, the same and on crosscompiler, and on native , so by adding that to cmake i was in hope cmake will just use those ones. But probably it check for "uname" or something firstly ?

**Comment by ptitSeb** [comment]
strange, I tested on my (using `ccmake`, the curse client of cmake, as I find it more conveniant) and it was looking for ppc-amigaos-gcc.
What version of cmake do you have?

**Comment by kas1e** [comment]
3.3.2

Not that its very important (as i anyway use pure bash script with necessary to build objects), but as you add  amigaos4 to cmake, maybe it worth to make it working :) 

I also have 2 cmake out logs, in one about Compiler: /usr/bin/c , and in another "The system is: CYGWIN - 2.3.1(0.291/5/3) - i686". Maybe, it still somehow think that it linux (cmake somewhere deep maybe do cygwin = linux), and then cmake fail.

**Comment by kas1e** [comment]
I also recieve new fixed driver. Through, we not sure if it works as it should , as test case still crasehes :( but on some simple tests looks like ok. Through in your example you say that it should be:

 uniform #3 : "_gl4es_FrontMaterial.emission" (builtin)  type=GL_FLOAT_VEC4 size=1
 uniform #4 : "_gl4es_FrontMaterial.ambient" (builtin)  type=GL_FLOAT_VEC4 size=1

While in my new debug output its a bit shifted, i.e.:

 uniform #0 : "_gl4es_NormalMatrix" (builtin)  type=GL_FLOAT_MAT3 size=1
 uniform #1 : "_gl4es_ModelViewProjectionMatrix" (builtin)  type=GL_FLOAT_MAT4 size=1
 uniform #2 : "_gl4es_FrontMaterial.emission" (builtin)  type=GL_FLOAT_VEC4 size=1
 uniform #3 : "_gl4es_FrontMaterial.ambient" (builtin)  type=GL_FLOAT_VEC4 size=1
 uniform #4 : "_gl4es_FrontMaterial.diffuse" (builtin)  type=GL_FLOAT_VEC4 size=1
 uniform #5 : "_gl4es_FrontMaterial.specular" (builtin)  type=GL_FLOAT_VEC4 

I.e. #3 ambient, #4 diffuse , like shifted on 1 position , but maybe that ok and it was just typo in your example ?

Anyway, test-case still crashes :( There is new debug output:
http://kas1e.mikendezign.com/aos4/gl4es/tests/fixed_structs_debug.txt

**Comment by kas1e** [comment]
Also with enabled -DDEBUG=1, latest texture.c didn't compiles, as well as glx.c on linking stage bring undef to "create_rpi_window". All other objects compiles fine

**Comment by ptitSeb** [comment]
The new driver seems to work fine. I don't think the shifted position is a problem. 

The log seems fine now to me, I don't see anything wrong. How is the rendering before the crash? is it correct

(I'll try to fix the build error with texture.c and glx.c)

**Comment by ptitSeb** [comment]
I don't have error on `texture.c` with DEBUG=1 on my side. What error do you have ?

**Comment by kas1e** [comment]
Its even didn't render anything, just open window and then baam :)
If i put a lot of Delays() around every function,  then i can see that it tries to render something .. something not related to the cube at all. Just some strange triangle, like everything wrong..

I start to fear it may be issue with our shaders, like it compiles, run , but just works wron.

What you think if we will try to change
ss = (nVP>0. && lVP>0.)?(_gl4es_FrontLightProduct_0.specular.xyz):vec3(0.);
on some dummy value ? 

Or, maybe we can try to shorted vertex shader, just to see, if it shader's bug at all or not..

As for texture error with DDEBUG:
src/gl/texture.c:2518: error: 'glstate_t' has no member named 'bound_stream'

**Comment by ptitSeb** [comment]
If you want, just cut the lightning: comment the `glEnable(GL_LIGHTNING);` that will greatly simplify the shader...

**Comment by kas1e** [comment]
Yeah, disabled in crash code glEnable(GL_LIGHTING); and still lockup ! 
debug: http://kas1e.narod.ru/no_light.txt

And shaders indeed very simple ones (imho).

**Comment by ptitSeb** [comment]
So I guess this is not the shader, as it's the same as last time.

How is the rendering this time? Still grabage?

**Comment by ptitSeb** [comment]
Also, can you comment those lines:
```
			glDrawBuffer(GL_BACK);
			display();
			glDrawBuffer(GL_FRONT);

```
that are after the SwapBuffers, just to simplify the code and try to pinpoint the piece that cause the freeze.

**Comment by kas1e** [comment]
Ops sorry i was wrong : No crash ! Cube rotate without gl_Enable(GL_LIGHTING);

Wtf..

**Comment by ptitSeb** [comment]
Ah, so it can be the a shader issue.

**Comment by kas1e** [comment]
Yeah, those ones commented already, i just didn't update test case, there is new src:
http://kas1e.narod.ru/gl4es_crash.c

**Comment by kas1e** [comment]
Ok.. now probably need to find what in shader cause that, so i can then report it to shader's code author (damn!)

**Comment by ptitSeb** [comment]
Can the guy that develop the ogles2 driver help debugging here?

Now, if you want to simplify the shader yourself, you will need to hack-in in `src/gl/fpe_shader.c`
It's mainlys mainy `sprintf`, so you will have to comment inside the string, to comment in the shader (it will be better then commenting the code generation itself I think).

You are interested in lines 354, 372 & 391 mainly.

**Comment by kas1e** [comment]
> Can the guy that develop the ogles2 driver help debugging here?

Yeah he already, just its 2 different persons. One develop ogles2.libary (which based on warp3d), and another develop warp3d (in where shaders compiler placed). First one fast on responce and stuff, second one not so :(

**Comment by kas1e** [comment]
Maybe for begining trying to put some dummy value to "ss =" , like "ss = 1" ..  as it look like string which may cause issue ..

**Comment by ptitSeb** [comment]
Just comment, like `sprintf(buff, "//ss = (nVP>0. && lVP>0.)?(%s%d.specular.xyz):vec3(0.);\n", fm_specular, i);` that would be the simpler...

**Comment by ptitSeb** [comment]
That makes a lots of emulatio layer when you look at it!

Program ---`OpenGL`---> GL4ES ---`GLES2`--->OGLES2 driver---`whatever it is`--->Warp3D-->Hardware

**Comment by kas1e** [comment]
Yeah :) But works fast enough already :)

Checked commenting out stuff, and , if i just comment out line 354 , then no crash, but no crash because shader fail, and nothing renders:

[quote]
glGetShaderInfoLog(256, 1000, 0x0, 0x6c3ab568)
LIBGL: FPE Vertex shader compile failed: ERROR: Code generation failed.
Error log:
INTERNAL ERROR: Variable is missing its CGVars:
	       117: OpLoad: FloatVector3: tmp117 << aa
INTERNAL ERROR: I/FAdd instruction's source parameters are missing
INTERNAL ERROR: I/FAdd instruction's source parameters are missing
INTERNAL ERROR: (Vector/Matrix)TimesScalar instruction's source parameters are missing
INTERNAL ERROR: I/FAdd instruction's source parameters are missing
INTERNAL ERROR: Dot instruction's source parameters are missing
INTERNAL ERROR: Store instruction's source parameter is incomplete:
	            OpStore: : tmp127 >> Color 
[/quote]

**Comment by ptitSeb** [comment]
Mmmm,
 then maybe you have to affect a bummy value indeed:
try: `sprintf(buff, "aa = 1.0;//%s.xyz * _gl4es_LightSource_%d.ambient.xyz;\n", fm_ambient, i);`

**Comment by kas1e** [comment]
With that one fail as well:

ERROR: 76:1: 'assign' : cannot convert from ' const float' to ' temp highp 3-component vector of float'

**Comment by ptitSeb** [comment]
Ah oops, yes, `aa`, `dd` and `ss` are all vec3 (it's the 3 componant of the light/material interaction: ambient, diffuse and specular, and all are a color)
so use `sprintf(buff, "aa = vec3(1.0);//%s.xyz * _gl4es_LightSource_%d.ambient.xyz;\n", fm_ambient, i);`

**Comment by kas1e** [comment]
So, only original "aa =" is ok , "dd =" and "ss =" , both cause lockup if they not dummy but originals (together , or per one, does not matter).

**Comment by kas1e** [comment]
Probably our shaders dind't like those "?" constructs seems so .. Is there any fast way we can check if it the cause ? I mean somehow rephrase those strings, without having "?" in them ?

**Comment by ptitSeb** [comment]
So maybe it's the ternary operator that ogles2 or warp3d doesn't like.
Try replacing it by simple `if/then/else`
like:
line 354 `sprintf(buff, "if(nVP>0.) dd=(nVP * %s%d.diffuse.xyz); else dd=vec3(0.);\n", fm_diffuse, i);`
line 391 `sprintf(buff, "if(nVP>0. && lVP>0.) ss=(%s%d.specular.xyz); else ss=vec3(0.);\n", fm_specular, i);`

**Comment by ptitSeb** [comment]
same idea ;)
(I'm off to bed now)

**Comment by kas1e** [comment]
Damn, same crash :( 

Will check more deeply tomorrow. 

Thanks for help !

**Comment by ptitSeb** [comment]
You can also remove the `else` part:
line 354 `sprintf(buff, "dd=vec3(0.0);\nif(nVP>0.) dd=(nVP * %s%d.diffuse.xyz);\n", fm_diffuse, i);
line 391 sprintf(buff, "ss=vec3(0.0);\nif(nVP>0. && lVP>0.) ss=(%s%d.specular.xyz);\n", fm_specular, i);

**Comment by kas1e** [comment]
still same :( 

Probably need to throw that shader on the author of shaders.. Damn, strange enough !

**Comment by ptitSeb** [comment]
A last test you can do is remove completly the `if` (of course rendering will not be correct).

line 354 `sprintf(buff, "dd=vec3(0.0);\ndd=(nVP * %s%d.diffuse.xyz);\n", fm_diffuse, i);`
line 391 `sprintf(buff, "ss=vec3(0.0);\nss=(%s%d.specular.xyz);\n", fm_specular, i);`

**Comment by kas1e** [comment]
With those lines it works ! Renders better, but as you say uncorrectly. I can see rotated cube , can see some green and red normals, but others normal are black. But that expected probably.

I also have an answer yesterday from the developer who works on shaders, that what he say:

-----
Before I start looking for a shader compiler bug, I'd like to know that it actually gets that far. As in, is the shader being compiled successfully all the way through to Warp3D ? Or, is something failing beforehand? For example, it's a larger shader and the lines in question are rather long. Is the buffer that the shader is sprintf()'ing to big enough?

The shader compiler's if/else functionality is fairly well tested, and those if/else sections are pretty basic. Plus, the vertex shader compiles just fine when I do it manually.
-----

Probably he just didn't check our debug logs where all looks fine enough..

**Comment by kas1e** [comment]
Interesting is that for our last checks without IF , we have generate those lines:

dd = vec3(0.0);
dd =(nVP * gl_Color.xyz *_gl4es_LightSource_0.diffuse.xyz);

and

ss = vec3(0.0);
ss = (_gl4es_FrontLightProduct_0.specular.xyz);

Is they correct ? I am mostly about ss = (_gl4es_FrontLightProduct_0.specular.xyz); , as in sprinfs i see %s%d,

**Comment by kas1e** [comment]
Also what i note (dunno if it related anyhow)

Is that with line:
line 354 sprintf(buff, "dd=vec3(0.0);\ndd=(nVP * %s%d.diffuse.xyz);\n", fm_diffuse, i);

We have in log:
dd =(nVP * gl_Color.xyz *_gl4es_LightSource_0.diffuse.xyz);

While with line 
ine 354 `sprintf(buff, "dd=vec3(0.0);\nif(nVP>0.) dd=(nVP * %s%d.diffuse.xyz);\n", fm_diffuse, i);

We have:
if(nVP>0.) dd = (nVP * gl4es_Color.xyz *_gl4es_LightSource_0.diffuse.xyz);


I.e. see gl_Color and gl4es_Color differences . It should be like this ?

**Comment by kas1e** [comment]
Well face alarm, just as both shaders visibly , i watch in another place. All fine in that terms ..

**Comment by ptitSeb** [comment]
the sprintf is `sprintf(buff, "ss=vec3(0.0);\nss=(%s%d.specular.xyz);\n", fm_specular, i);`, it's coherent with the output you see (`fm_specular`  is "_gl4es_FrontLightProduct_" in that case, and `i` is light `0`).

About the comment from Warp3D developper:
`buff` is a `1024` characters buffer, so no buffer overflow here. This code has been tested a lot, also using valgrind on linux, I'm pretty sure there is no buffer overflow.

Also, the issue is not at compile time it seems, but really at run time, with some corruption happening at run time.
Maybe it's the combination of vec3 and if?

Can you try something slightly different:
line 354 `sprintf(buff, "dd=dd=(nVP * %s%d.diffuse.xyz);\n", fm_diffuse, i)*((nVP>0.)?1.:0.);`
line 391 `sprintf(buff, "ss=(%s%d.specular.xyz);\n", fm_specular, i)*((nVP>0. && lVP>0.)?1.:0.);`

**Comment by kas1e** [comment]
Fail to compile with:

src/gl/fpe_shader.c:387: error: 'nVP' undeclared (first use in this function)
src/gl/fpe_shader.c:387: error: (Each undeclared identifier is reported only once
src/gl/fpe_shader.c:387: error: for each function it appears in.)
src/gl/fpe_shader.c:413: error: 'lVP' undeclared (first use in this function)

**Comment by kas1e** [comment]
(don't see that there 387 and 413 lines, just i keep previous ones commented out, so to be able fast check them back)

**Comment by ptitSeb** [comment]
Ah of course, I messed up things.

line 354 `sprintf(buff, "dd=(nVP * %s%d.diffuse.xyz)*((nVP>0.)?1.:0.);\n", fm_diffuse, i);`
line 391 `sprintf(buff, "ss=(%s%d.specular.xyz)*((nVP>0. && lVP>0.)?1.:0.);\n", fm_specular, i);`

Should be better that way

**Comment by kas1e** [comment]
With that one shader compilation fail:

glCompileShader(256)
glGetShaderiv(256, 0x8B81, 0x67d71564)
glGetShaderInfoLog(256, 1000, 0x0, 0x67d71568)
LIBGL: FPE Vertex shader compile failed: ERROR: Code generation failed.
Error log:
OpSelect instruction isn't implemented yet.
INTERNAL ERROR: (Vector/Matrix)TimesScalar instruction's source parameters are missing
OpSelect instruction isn't implemented yet.
INTERNAL ERROR: (Vector/Matrix)TimesScalar instruction's source parameters are missing
INTERNAL ERROR: I/FAdd instruction's source parameters are missing
INTERNAL ERROR: I/FAdd instruction's source parameters are missing
INTERNAL ERROR: (Vector/Matrix)TimesScalar instruction's source parameters are missing
INTERNAL ERROR: I/FAdd instruction's source parameters are missing
INTERNAL ERROR: Dot instruction's source parameters are missing
INTERNAL ERROR: Store instruction's source parameter is incomplete:
	            OpStore: : tmp122 >> Color

**Comment by ptitSeb** [comment]
Mmmm, try this then.

line 354 `sprintf(buff, "spot=(nVP>0.)?1.0:0.0;\ndd=(nVP * %s%d.diffuse.xyz)*spot;\n", fm_diffuse, i);`
line 391 `sprintf(buff, "spot=(nVP>0. && lVP>0.)?1.0:0.0;\nss=(%s%d.specular.xyz)*spot;\n", fm_specular, i);`

**Comment by kas1e** [comment]
That one seems better, but still fail:

LIBGL: FPE Vertex shader compile failed: ERROR: Code generation failed.
Error log:
OpSelect instruction isn't implemented yet.
INTERNAL ERROR: Store instruction's source parameter is incomplete:
	            OpStore: : tmp76 >> spot
OpSelect instruction isn't implemented yet.
INTERNAL ERROR: Store instruction's source parameter is incomplete:
	            OpStore: : tmp103 >> spot

**Comment by ptitSeb** [comment]
mmm, I'm not sure I understand the error. The code seems fine to me (if you try to check the shaders, you have error?)
Anyway, let's try with simple `if` and no ternary operator:
line 354 `sprintf(buff, "spot=0.0;\nif(nVP>0.) spot=1.0;\ndd=(nVP * %s%d.diffuse.xyz)*spot;\n", fm_diffuse, i);`
line 391 `sprintf(buff, "spot=0.0;\nif(nVP>0. && lVP>0.) spot=1.0;\nss=(%s%d.specular.xyz)*spot;\n", fm_specular, i);`

**Comment by kas1e** [comment]
With last ones shaders compiles fine, but then crashes :) 

Seems once "if" in the palce, then we crashes..

**Comment by kas1e** [comment]
Once i remove from last example that "if" string, i.e. keep just only:

spot=0.0;
dd=(nVP * _gl4es_Color.xyz * _gl4es_LightSource_0.diffuse.xyz)*spot; 

and

spot=0.0;
ss=(_gl4es_FrontLightProduct_0.specular.xyz)*spot; 

Then no crash ! Cube rotates ! Wtf ..

**Comment by ptitSeb** [comment]
Here, a version with no `if` or ternary at all! :)

line 354 `sprintf(buff, "spot=max(1.0, sign(nVP)+1);\ndd=(nVP * %s%d.diffuse.xyz)*spot;\n", fm_diffuse, i);`
line 391 `sprintf(buff, "ss=(%s%d.specular.xyz)*spot;\n", fm_specular, i)*spot*max(1.0,sign(lVP)+1.0);`

And this shader should be exact (just slower then original... but not that much).

**Comment by ptitSeb** [comment]
Or this one should works too:

line 354 `sprintf(buff, "spot=step(0.0, nVP);\ndd=(nVP * %s%d.diffuse.xyz)*spot;\n", fm_diffuse, i);`
line 391 `sprintf(buff, "ss=(%s%d.specular.xyz)*spot;\n", fm_specular, i)*spot*step(0.0,lVP);`

**Comment by kas1e** [comment]
In both cases line about "ss=" , fails:

src/gl/fpe_shader.c: In function 'fpe_VertexShader':
src/gl/fpe_shader.c:429: error: 'spot' undeclared (first use in this function)
src/gl/fpe_shader.c:429: error: (Each undeclared identifier is reported only once
src/gl/fpe_shader.c:429: error: for each function it appears in.)
src/gl/fpe_shader.c:429: warning: implicit declaration of function 'step'
src/gl/fpe_shader.c:429: error: 'lVP' undeclared (first use in this function)

**Comment by kas1e** [comment]
In previous version shaders fail to compile with words : about wrong operand ypes: no operation "+" exists that takes a left hand operatn on tupe glogal highp float" , etc

But last version works fine :)

**Comment by kas1e** [comment]
So .. what did it mean at all , that once we have "if" it all crashes in runtime (shader's bug).
As well as if we have "?" betwen pairs, crashes too ? (but probably that the same "if" in end, so same code)

**Comment by kas1e** [comment]
And i notice that something wrong with colors happens as well with last string (on of normales have color of backgorund, while should be purple).

But that not so matter anyway, for test it was enough. If that it shader's issue it should be fixed in our shader compiller.

**Comment by ptitSeb** [comment]
Yeah, the ternary operator and if probably share the same code in warp3d shaders...
Now that you have a version of the shader that work (with `step`), and a version that doesn't (with `if` or with `?`), you can open a bug in Warp3D side...

Now, I'll see if I can add an optional code path in fpr_shader to limit the ternary operator in the vertex shader, at least for lights.
Maybe it can be faster on the Pandora, I have to test.

**Comment by kas1e** [comment]
Yeah, but only make it optional / possible to remove , as those problems should be fixed in drivers of coruse, or it will lead to massive problems when i will try to port something a little more heavy :)

**Comment by ptitSeb** [comment]
@kas1e with commit 077d870a7c1ddcc57190f1a4c7a8586ad49e32a2 vertex shaders generated by `fpe_shader.c` will have no ternary operator.

**Comment by kas1e** [comment]
And that commit works, yep. Let it be here then until fix for shader's compiler is come.

But now, i found another issues, that can be easy related to Endianes, or, that can be related to the replacement of ternary operator. 

There is 2 screenshots, one i catch with our 1.1 version of opengl  (called MiniGL).
Another one, its from GL4ES i.e. what we test.

MiniGL one:  http://kas1e.mikendezign.com/aos4/gl4es/tests/mgl_rotate.jpg
GL4ES one: http://kas1e.mikendezign.com/aos4/gl4es/tests/gl4es_rotate.jpg

I catch screenshots on "the same" place, so, we can see that colors of one of the sides of the cube are different and not as they expected to be. In minigl you can see its purple , in the gl4es vesion that side are just of the same as background color. 

Can  you plz test , if you have on Pandora that "purple" normale with ternary and without ternary operators, or you also have one of the sides of cube with the same as background color.

Thanks !

**Comment by ptitSeb** [comment]
I have purple with the shader with ternary operator.

I'll rebuild without and retest soon.

**Comment by ptitSeb** [comment]
I also have purple with the alternate shader. So it's something else.

**Comment by kas1e** [comment]
Hmm.. I think about Endianes, but usually with endianes all colors wrong, not only one side.. Another idea , is can it be that uniform structures, somehow "swapped" somewhere, so it can cause such effect ?

**Comment by ptitSeb** [comment]
Maybe yes. But colors will be an attribute, not a uniform here. can you create a full log?

**Comment by kas1e** [comment]
Here is: http://kas1e.mikendezign.com/aos4/gl4es/tests/wrong_color.txt

Through its 5mb of size, its very fast fills by data , and i do not edit it just in case something intersting will be at end on somewhere.

**Comment by ptitSeb** [comment]
The log looks fine to me. Maybe one of the driver (OGLES or Warp3D) doesn't like the fact that the Color attribute (and the Normal too, but it's not really visible) is a "constant" attribute. It would be interresting to make it a non-constant one.
In your sample, change:
```
glColor3fv(&colors[i][0]);
```
by
```
    GLfloat col[8][3];
    for (int j=0; j<8; j++) memcpy(col[j], colors[i], 3*4);
    glEnableClientState(GL_COLOR_ARRAY);
    glColorPointer(3, GL_FLOAT, 0, col);
```
(yeah ugly code, it's just for testing :p )

**Comment by kas1e** [comment]
And yes, everything good with that changed code.. DAMN ! Sorry that i throw it all at you, like you need to fix all our bugs :) Can we reduce the test somehow so we can be sure what wrong ?

**Comment by kas1e** [comment]
Btw, also it fix "geometry" : if you check previous 2 screenshots, you can see that right/down area in minigl version are "blue" , while, when we run it with glColor3fv(&colors[i][0]); on gl4es, then there is no that blue area. What mean its not only color, but normales too ?

**Comment by ptitSeb** [comment]
The Normal are set the same way the colors were. So I guess there are no setup properly. But with this zoomed cube, I'm not sure you see the effect of a bad normal correctly (the reflection should be slightly different then with minigl, the moment where the face is "white" because it completly reflect the light).

**Comment by ptitSeb** [comment]
So I think you can open a new ticket for the Attribute issue:
Vertex Attribute set with `glVertexAttrib4fv()` does not seems to be correctly updated.

**Comment by ptitSeb** [comment]
I prefer to not try to compensate for this bug for now. It is possible, but that mean creating new array, so it's really a waste and prefer to avoid that (unless there is no chance to get a fix).

**Comment by kas1e** [comment]
Sure, fix in driver are must. Will write to author now.

Thanks!

**Comment by kas1e** [comment]
@ptitSeb
By the way, what you use on Pandora when build pure GLES2 examples ? SDL or EGL ?

**Comment by ptitSeb** [comment]
I use SDL, it's much easier to setup then EGL directly.

**Comment by kas1e** [comment]
That cool ! I do not know if it will be too much to ask , but as our shader's compiler author "want test case", is it possible for you to create a GLES2 example from that test-case debug output ? I mean i can see in debug log what functions calls, etc, but i do not know how to make it all works on GLES2 code. All i know that i need to take those 2 shaders, load them, and then mimic logic of what i do in test case .  But while on GL it looks easy enough, on GLES2 it can be some mess ?:)

**Comment by ptitSeb** [comment]
I'll see what I can do.

**Comment by ptitSeb** [comment]
But I'm not sure why he cannot used the current version that use gl4es. He wants to change the shaders?

**Comment by kas1e** [comment]
He want to shorten all code and layers in between to minimum, so can find where problem come. For first probably it will be pure gles2, then probable he will eleminate gles2 and made direct warp3d sample.

**Comment by ptitSeb** [comment]
Well, I'll see what I can do, it's a bit of work an hour or two to get something working I guess.

**Comment by kas1e** [comment]
@ptitSeb
Can you also give me your paypal address, want to motivate you a bit by some donation :) If you doesnt mind of course

**Comment by ptitSeb** [comment]
I assume SDL2 is ok, right?

I do have a paypal yes: https://www.paypal.com/donate/?token=RUuvxCN8rmpvaup-t30Qd-uuGmLvTDcEc2jR52jgOnE5Qctz5qC8CcZs0n9NeLYGozQPYG

**Comment by kas1e** [comment]
Yep, sdl2 good

**Comment by kas1e** [comment]
@ptitSeb

Recieved fix in ogles2.library for that issues about Vertex Attribute set with glVertexAttrib4fv() does not seems to be correctly updated.  

As author says, it  redundant vertex-data detector had a bug that eventually caused "one-for-all" vertex-attributes like those defined via glVertexAttrib4f not being updated because other parts of the lib would believe that the data did not change; the old value would then be used again unless another VA change eventually caused full  invalidation of the internal VA state.

And he thank you for help :)

ps. check pp, some little motivator, hope not the last one :)

**Comment by ptitSeb** [comment]
Good it's fixed!

(In fact gl4es is very similar to the OGES2 driver, so yeah, I understand well the issue, and gl4es implement similar mecanism).

(Im' almost done with the pura ogles2 sample, almost...)

**Comment by ptitSeb** [comment]
@kas1e : here is you ogles2 test: [test.c](ptitseb.openpandora.org/test.c)

It's based on SDL2, and needs `-lm` at least, and the OGLES2 library.

It mimics what the GL sample do.
As you can see, gl4es handle a lot of thing behind the scene...

**Comment by kas1e** [comment]
Thanks a bunch ! Hope we will be able to trace problem with shaders now 

Thanks!

**Comment by kas1e** [comment]
And yep, GLES version looks much harder than GL ones to say truth :)

**Comment by ptitSeb** [comment]
Also, I think we can close this ticket now.
The crash/freeze has a workaround in gl4es, you have way to reproduce to open a ticket on Warp3D side, and the single Vertex Attrib bug in OGLES2 driver is also fixed, you can continue your tests now (with some texture, so you can open tickets in gl4es for not handling BigEndian stuff correctly :p )

**Comment by ptitSeb** [comment]
(And thanks for the donation :) ! )

**Comment by kas1e** [comment]
Yeah, let's close that one , will cookie texture based now :)
